`make sphinx` Breaks when Certain Untracked Files are Present

* `make sphinx` fails if I have some file like "pep-0684-old.rst" (the filename regex in pep_sphinx_extensions/pep_zero_generator/pep_index_generator.py isn't precise enough)
* `make sphinx` fails if I have a venv dir named "venv-spam" sitting at the top of the repo (tries to parse a vendor.txt file inside the venv)

In both cases I'd expect it to work fine, ignoring the relevant files.

I *think* the issue I'm seeing trying to build locally is this one (rather than #2402), since the linting works fine, but the rendering is complaining about old files inside the `.git` folder:
```
... snip ...
checking consistency... /home/ncoghlan/devel/peps/.git/logs/refs/heads/augmented-assignment-expressions.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/.git/logs/refs/remotes/pr/augmented-assignment-expressions.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/.git/refs/heads/augmented-assignment-expressions.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/.git/refs/remotes/pr/augmented-assignment-expressions.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/docs/build.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/docs/rendering_system.rst: WARNING: document isn't included in any toctree
/home/ncoghlan/devel/peps/pep-0593.rst:265: WARNING: Citation [struct-doc] is not referenced.
/home/ncoghlan/devel/peps/pep-0553.rst:283: WARNING: Citation [envar] is not referenced.
/home/ncoghlan/devel/peps/pep-0554.rst:1645: WARNING: Citation [CSP] is not referenced.
/home/ncoghlan/devel/peps/pep-0554.rst:1677: WARNING: Citation [mp-conn] is not referenced.
/home/ncoghlan/devel/peps/pep-0554.rst:1686: WARNING: Citation [main-thread] is not referenced.
/home/ncoghlan/devel/peps/pep-0554.rst:1709: WARNING: Citation [nathaniel-asyncio] is not referenced.
/home/ncoghlan/devel/peps/pep-0554.rst:1712: WARNING: Citation [extension-docs] is not referenced.
/home/ncoghlan/devel/peps/pep-0520.txt:416: WARNING: Citation [nick_concern] is not referenced.
/home/ncoghlan/devel/peps/pep-0520.txt:419: WARNING: Citation [orig] is not referenced.
/home/ncoghlan/devel/peps/pep-0520.txt:422: WARNING: Citation [followup1] is not referenced.
/home/ncoghlan/devel/peps/pep-0520.txt:425: WARNING: Citation [followup2] is not referenced.
/home/ncoghlan/devel/peps/pep-0484.txt:2490: WARNING: Citation [peps] is not referenced.
done
preparing documents... done
Traceback (most recent call last):
  File "/home/ncoghlan/devel/peps/build.py", line 86, in <module>
    app.build()
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/application.py", line 337, in build
    self.builder.build_update()
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/builders/__init__.py", line 294, in build_update
    self.build(to_build,
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/builders/__init__.py", line 358, in build
    self.write(docnames, list(updated_docnames), method)
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/builders/__init__.py", line 529, in write
    self._write_parallel(sorted(docnames),
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/builders/__init__.py", line 556, in _write_parallel
    self.write_doc(firstname, doctree)
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/sphinx/builders/html/__init__.py", line 631, in write_doc
    ctx = self.get_doc_context(docname, body, metatags)
  File "/home/ncoghlan/devel/peps/pep_sphinx_extensions/pep_processor/html/pep_html_builder.py", line 39, in get_doc_context
    if len(toc_tree[0]) > 1:
  File "/home/ncoghlan/devel/peps/.venv/lib64/python3.10/site-packages/docutils/nodes.py", line 654, in __getitem__
    return self.children[key]
IndexError: list index out of range
make: *** [Makefile:9: render] Error 1
```

I'll try the #2415 branch to see if that resolves the issue.

--- a/conf.py
+++ b/conf.py
@@ -8,6 +8,11 @@ import sys
 
 sys.path.append(str(Path("pep_sphinx_extensions").absolute()))
 
+# Add 'include_patterns' as a config variable
+from sphinx.config import Config
+Config.config_values['include_patterns'] = [], 'env', []
+del Config
+
 # -- Project information -----------------------------------------------------
 
 project = "PEPs"
@@ -25,24 +30,19 @@ source_suffix = {
 }
 
 # List of patterns (relative to source dir) to ignore when looking for source files.
+include_patterns = [
+    # Required for Sphinx
+    "contents.rst",
+    # PEP files
+    "pep-????.???",
+    # PEP ancillary files
+    "pep-????/*.rst",
+    # Documentation
+    "docs/*.rst",
+]
 exclude_patterns = [
-    # Windows:
-    "Thumbs.db",
-    ".DS_Store",
-    # Python:
-    ".venv",
-    "venv",
-    "requirements.txt",
-    # Sphinx:
-    "build",
-    "output.txt",  # Link-check output
-    # PEPs:
-    "pep-0012",
-    "README.rst",
-    "CONTRIBUTING.rst",
-    "pep_sphinx_extensions/LICENCE.rst",
-    # Miscellaneous
-    ".codespell",
+    # PEP Template
+    "pep-0012/pep-NNNN.rst",
 ]
 
 # -- Options for HTML output -------------------------------------------------

--- a/conf.py
+++ b/conf.py
@@ -34,7 +34,8 @@ include_patterns = [
     # Required for Sphinx
     "contents.rst",
     # PEP files
-    "pep-????.???",
+    "pep-????.rst",
+    "pep-????.txt",
     # PEP ancillary files
     "pep-????/*.rst",
     # Documentation

--- a/pep_sphinx_extensions/__init__.py
+++ b/pep_sphinx_extensions/__init__.py
@@ -6,6 +6,7 @@ from typing import TYPE_CHECKING
 
 from docutils.writers.html5_polyglot import HTMLTranslator
 from sphinx import environment
+from sphinx import project
 
 from pep_sphinx_extensions.pep_processor.html import pep_html_builder
 from pep_sphinx_extensions.pep_processor.html import pep_html_translator
@@ -16,6 +17,37 @@ from pep_sphinx_extensions.pep_zero_generator.pep_index_generator import create_
 
 if TYPE_CHECKING:
     from sphinx.application import Sphinx
+    from sphinx.config import Config
+
+
+def find_files(self: environment.BuildEnvironment, config: Config, _b) -> None:
+    """Find all pep source files."""
+    import fnmatch
+    from pathlib import Path
+
+    root = Path(self.project.srcdir).absolute()
+    self.project.docnames = set()
+    for pattern in config.include_patterns:
+        for path in root.glob(pattern):
+            filename = str(path.relative_to(root))
+            if any(fnmatch.fnmatch(filename, pattern) for pattern in config.exclude_patterns):
+                continue
+
+            doc_name = self.project.path2doc(filename)
+            if not doc_name:
+                continue
+
+            if doc_name not in self.project.docnames:
+                self.project.docnames.add(doc_name)
+                continue
+
+            other_files = [str(f.relative_to(root)) for f in root.glob(f"{doc_name}.*")]
+            project.logger.warning(
+                f'multiple files found for the document "{doc_name}": {other_files!r}\n'
+                f'Use {self.doc2path(doc_name)!r} for the build.', once=True)
+
+
+environment.BuildEnvironment.find_files = find_files
 
 
 def _depart_maths():

